tests/e2e/vmc_upgrade_tkg.go

```diff
@@ -0,0 +1,275 @@
+		gomega.Expect(len(ssPodsBeforeScaleUp.Items) == int(replicas)).To(gomega.BeTrue(),
+			"Number of Pods in the statefulset should match with number of replicas")
+
+		// Verify volumes are attached to the nodes.
+		for _, sspod := range ssPodsBeforeScaleUp.Items {
+			_, err := client.CoreV1().Pods(namespace).Get(ctx, sspod.Name, metav1.GetOptions{})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			for _, volumespec := range sspod.Spec.Volumes {
+				if volumespec.PersistentVolumeClaim != nil {
+					pv := getPvFromClaim(client, statefulset.Namespace, volumespec.PersistentVolumeClaim.ClaimName)
+					ginkgo.By("Verify CnsNodeVmAttachment CRD is created")
+					volumeID := pv.Spec.CSI.VolumeHandle
+					// svcPVCName refers to PVC Name in the supervisor cluster.
+					svcPVCName := volumeID
+					volumeID = getVolumeIDFromSupervisorCluster(svcPVCName)
+					gomega.Expect(volumeID).NotTo(gomega.BeEmpty())
+					verifyCRDInSupervisor(ctx, f, sspod.Spec.NodeName+"-"+svcPVCName,
+						crdCNSNodeVMAttachment, crdVersion, crdGroup, true)
+
+					ginkgo.By(fmt.Sprintf("Verify volume: %s is attached to the node: %s",
+						pv.Spec.CSI.VolumeHandle, sspod.Spec.NodeName))
+					var vmUUID string
+					vmUUID, err = getVMUUIDFromNodeName(sspod.Spec.NodeName)
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					isDiskAttached, err := e2eVSphere.isVolumeAttachedToVM(client, volumeID, vmUUID)
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					gomega.Expect(isDiskAttached).To(gomega.BeTrue(),
+						fmt.Sprintf("Volume is not attached to the node, %s", vmUUID))
+
+					ginkgo.By(fmt.Sprintf("Verify volume: %s is attached to the node: %s",
+						pv.Spec.CSI.VolumeHandle, sspod.Spec.NodeName))
+					verifyIsAttachedInSupervisor(ctx, f, sspod.Spec.NodeName+"-"+pv.Spec.CSI.VolumeHandle,
+						crdVersion, crdGroup)
+
+				}
+			}
+		}
+
+		ginkgo.By("Upgrade TKG")
+		if vmcUser == devopsUser {
+			ginkgo.By("Get WCP session id")
+			gomega.Expect((e2eVSphere.Config.Global.VmcDevopsUser)).NotTo(gomega.BeEmpty(), "Devops user is not set")
+			wcpToken = getWCPSessionId(vmcWcpHost, e2eVSphere.Config.Global.VmcDevopsUser,
+				e2eVSphere.Config.Global.VmcDevopsPassword)
+			framework.Logf("vmcWcpHost %s", vmcWcpHost)
+		} else {
+			ginkgo.By("Get WCP session id")
+			gomega.Expect((e2eVSphere.Config.Global.VmcCloudUser)).NotTo(gomega.BeEmpty(), "Cloud user is not set")
+			framework.Logf("vmcWcpHost %s", vmcWcpHost)
+			wcpToken = getWCPSessionId(vmcWcpHost, e2eVSphere.Config.Global.VmcCloudUser,
+				e2eVSphere.Config.Global.VmcCloudPassword)
+		}
+		vmImage := getVMImages(vmcWcpHost, wcpToken)
+		size := len(vmImage.Items)
+
+		for i := 0; i < size; i++ {
+			framework.Logf(vmImage.Items[i].Spec.ProductInfo.Version)
+			if vmImage.Items[i].Spec.ProductInfo.Version == tkg_image {
+				flag = true
+				break
+			}
+		}
+
+		gomega.Expect(flag).NotTo(gomega.BeFalse(), "vmimage passed is not present in the cluster")
```

Comment:
we can do this at the start of the test before creating sts

Reply:
done